File: JS/src/translator/TAProgram.js
const StaticSymbolTable = require('./symbol/StaticSymbolTable')
const TAInstruction = require('./TAInstruction')
const TAInstructionType = require('./TAInstructionType')
const SymbolType = require('./symbol/SymbolType')

// Three Address
class TAProgram {
    constructor() {

        this.instructions = []
        this.labelCounter = 0
        this.staticSymbolTable = new StaticSymbolTable()
    }

    add(code) {
        this.instructions.push(code)
    }

    getInstructions() {
        return this.instructions
    }

    getStaticSymbolTable() {
        return this.staticSymbolTable
    }

    addLabel() {
        const label = "L" + this.labelCounter++
        const taCode = new TAInstruction(TAInstructionType.LABEL, null, null, null, null)
        taCode.setArg1(label)
        this.instructions.push(taCode)
        return taCode
    }

    setStaticSymbols(symbolTable) {
        for (const symbol of symbolTable.getSymbols()) {
            if (symbol.getType() == SymbolType.IMMEDIATE_SYMBOL) {
                this.staticSymbolTable.add(symbol);
            }
        }

        for (const child of symbolTable.getChildren()) {
            this.setStaticSymbols(child)
        }
    }
}
module.exports = TAProgram

File: Java/src/translator/TAProgram.java
package translator;

import org.apache.commons.lang.StringUtils;
import translator.symbol.StaticSymbolTable;
import translator.symbol.Symbol;
import translator.symbol.SymbolTable;
import translator.symbol.SymbolType;

import java.util.ArrayList;

/**
 * Author : lihuichuan
 * Time   : 2020/12/11
 */
public class TAProgram {
    private ArrayList<TAInstruction> instructions = new ArrayList<>();
    private int labelCounter = 0;
    private StaticSymbolTable staticSymbolTable = new StaticSymbolTable();

    public void add(TAInstruction code) {
        instructions.add(code);
    }

    public ArrayList<TAInstruction> getInstructions() {
        return instructions;
    }

    @Override
    public String toString() {
        ArrayList lines = new ArrayList<String>();
        for (TAInstruction opcode : instructions) {
            lines.add(opcode.toString());
        }
        return StringUtils.join(lines, "\n");
    }

    public void setStaticSymbols(SymbolTable symbolTable) {
        for (Symbol symbol : symbolTable.getSymbols()) {
            if (symbol.getType() == SymbolType.IMMEDIATE_SYMBOL) {
                staticSymbolTable.add(symbol);
            }
        }

        for (SymbolTable child : symbolTable.getChildren()) {
            setStaticSymbols(child);
        }
    }

    public StaticSymbolTable getStaticSymbolTable() {
        return this.staticSymbolTable;
    }

    public TAInstruction addLabel() {
        String label = "L" + labelCounter++;
        TAInstruction taCode = new TAInstruction(TAInstructionType.LABEL, null, null, null, null);
        taCode.setArg1(label);
        instructions.add(taCode);
        return taCode;
    }

}
